### PR TITLE
[SPARK-41585][YARN] The Spark exclude node functionality for YARN should work independently of dynamic allocation

### PR DESCRIPTION
**Pull request by LucaCanali**
### What changes were proposed in this pull request?
The Spark exclude node functionality for Spark on YARN, introduced in [SPARK-26688](https://issues.apache.org/jira/browse/SPARK-26688), allows users to specify a list of node names that are excluded from resource allocation. This is done using the configuration parameter: `spark.yarn.exclude.nodes`
The feature currently works only for executors allocated via dynamic allocation. To use the feature on Spark 3.3.1, for example, one may need also to configure `spark.dynamicAllocation.minExecutors=0` and `spark.executor.instances=0`, therefore relying on executor resource allocation only via dynamic allocation.

### Why are the changes needed?
This proposes to extend the use of Spark exclude node functionality for YARN beyond dynamic allocation, which I believe makes it more consistent also with what the documentation reports for this feature/configuration parameter.

### Does this PR introduce _any_ user-facing change?
Yes, this allows using the executor exclude nodes feature for Spark on YARN also when not using dynamical allocation.

### How was this patch tested?
A unit test has been added for this + manual tests on a YARN cluster.

### CODE REVIEW

**Review by mridulm — COMMENTED** [review]
Looks reasonable, but I have not looked at this in a while.

+CC @tgravescs, @attilapiros, @xkrogen

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 1]
```diff
@@ -563,6 +565,7 @@ private[yarn] class YarnAllocator(
           }
         }
 
+        allocatorNodeHealthTracker.setSchedulerExcludedNodes(excludeNodes)
```
Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:568]: this feels like a weird way to do this based on the YarnAllocatorNodeHealthTracker already reads the excludeNode and has it as a private variable.  This api was meant for excluding nodes that had dynamic failures... ie the scheduler excluded them.  These are not excluded by the scheduler but by user manually.

what if we just have it call refreshExcludedNodes when the YarnAllocatorNodeHealthTracker  is created?  Unless you are trying to dynamically change this config which it wasn't really meant for.

**Review by tgravescs — COMMENTED** [review]
looks good, can you add in a test for this?

**Review by attilapiros — COMMENTED** [review]
I have tested it manually on a YARN cluster.

Started Spark with fix number of executors and with one excluded host like:
`--num-executors 3 --conf spark.yarn.exclude.nodes="bla" `

The log lines contained:
```
23/02/18 06:39:18 INFO  yarn.YarnAllocatorNodeHealthTracker: [main]: adding nodes to YARN application master's excluded node list: List(bla)
```

LGTM after a unit test is added!

**Comment by tgravescs** [comment]
I was thinking just basic unit test were you set the config and then create the YarnAllocatorNodeHealthTracker and validate it initialized them properly and called the amclient with expected values. Add comment about covering non dynamic allocation.

**Review by tgravescs — APPROVED** [review]

**Comment by LucaCanali** [comment]
Thanks @tgraves , @attilapiros and @mridulm for reviewing this.
I guess this is now ready to be merged to master?

**Comment by tgravescs** [comment]
merged to master

**Pull request closed without merging**